src/Authentication/Authentication/Common/GraphSession.cs

```diff
@@ -0,0 +1,128 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Graph.PowerShell.Authentication
+{
+    using System;
+    using System.Threading;
+    /// <summary>
+    /// The current <see cref="GraphSession"/>.
+    /// </summary>
+    public class GraphSession: IGraphSession
+    {
+        static GraphSession _instance;
+        static bool _initialized = false;
+        static ReaderWriterLockSlim sessionLock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
+        internal Guid _graphSessionId;
+
+        /// <summary>
+        /// Gets or Sets <see cref="IAuthContext"/>.
+        /// </summary>
+        public IAuthContext AuthContext { get; set; }
+        public static GraphSession Instance
+        {
+            get
+            {
+                try
+                {
+                    sessionLock.EnterReadLock();
+                    try
+                    {
+                        if (null == _instance)
+                        {
+                            throw new InvalidOperationException(ErrorConstants.Codes.SessionNotInitialized);
+                        }
+                        return _instance;
+                    }
+                    finally
+                    {
+                        sessionLock.ExitReadLock();
+                    }
+                }
+                catch (LockRecursionException lockException)
+                {
+                    throw new InvalidOperationException(ErrorConstants.Codes.SessionLockReadRecursion, lockException);
+                }
+                catch (ObjectDisposedException disposedException)
+                {
+                    throw new InvalidOperationException(ErrorConstants.Codes.SessionLockReadDisposed, disposedException);
+                }
+            }
+        }
+        public GraphSession()
+        {
+            _graphSessionId = Guid.NewGuid();
+        }
+
+        /// <summary>
+        /// Initialize <see cref="GraphSession"/>.
+        /// </summary>
+        /// <param name="instanceCreator">A func to create an instance.</param>
+        /// <param name="overwrite">If true, overwrite the current instance. Otherwise do not initialize.</param>
+        public static void Initialize(Func<GraphSession> instanceCreator, bool overwrite)
+        {
+            try
+            {
+                sessionLock.EnterWriteLock();
+                try
+                {
+                    if (overwrite || !_initialized)
```

Comment:
>overwrite [](start = 24, length = 9)

the pattern here if you don't set the flag but has an instance seems like silent failure which leaves code error prone.  It should probably throw an exception that it is not overwriting the session.

Reply:
The assumption made here is that `Initialize` should only be called once, and if one wants to overwrite the current instance, then they have to do so by explicitly calling its overload that accepts `bool overwrite`.

Reply:
your intent is right.  The issue is that if you do call twice you won't know that it did not give you back a new GraphSession and making that assumption can lead to further bugs in the callers code.  That's why I think it's better to throw and say you must explicitly overwrite so that assumption that it did modify cannot be made.